src/nrnpython/nrnpy_nrn.cpp

```diff
@@ -1493,29 +1493,28 @@ static PyObject* NPySecObj_connect_safe(NPySecObj* self, PyObject* args) {
 static PyObject* NPySecObj_insert(NPySecObj* self, PyObject* args) {
     CHECK_SEC_INVALID(self->sec_);
     char* tname;
-    PyObject *tpyobj, *tpyobj2;
     if (!PyArg_ParseTuple(args, "s", &tname)) {
         PyErr_Clear();
         // if called with an object that has an insert method, use that
+        PyObject* tpyobj;
         if (PyArg_ParseTuple(args, "O", &tpyobj)) {
-            Py_INCREF(tpyobj);
-            Py_INCREF((PyObject*) self);
-            tpyobj2 = PyObject_CallMethod(tpyobj, "insert", "O", (PyObject*) self);
-            Py_DECREF(tpyobj);
-            if (tpyobj2 == NULL) {
-                Py_DECREF((PyObject*) self);
+            auto _tpyobj_tracker = nb::borrow(tpyobj);
+            // tpyobj.insert reqs incref self. Borrow it so if it fails the ref count rolls-back
+            auto self_obj = nb::borrow((PyObject*) self);
+            auto tpyobj2 = nb::steal(PyObject_CallMethod(tpyobj, "insert", "O", self_obj.ptr()));
+            if (!tpyobj2.is_valid()) {
                 PyErr_Clear();
                 PyErr_SetString(
                     PyExc_TypeError,
                     "insert argument must be either a string or an object with an insert method");
-                return NULL;
+                return nullptr;
             }
-            Py_DECREF(tpyobj2);
-            return (PyObject*) self;
+            self_obj.inc_ref();  // We return the object, need INCREF again
+            return self_obj.release().ptr();
```

Comment:
Let's continue the discussion there. We're at reference count `+2` and I'm not convinced that that's the correct number. It could be `+0`, `+1` or `+2`.

Previous discussion: https://github.com/neuronsimulator/nrn/pull/3023#discussion_r1698486920

Reply:
This claims we need to return a "new reference":
https://docs.python.org/3/c-api/structures.html#c.PyCFunction

Hence, `+1` for returning a new reference (which is exactly what NRN currently does). That leaves whether we need `+2` because we call `PyObject_CallMethod` with `"insert"`.

Reply:
Ok, so key points to take from previous discussion:
 - we return `self`. It is expected that this functions doesn't take ownership of `self`. therefore +1
 - We pass `self` to `tpyobj, insert`. It is unclear whether the function internally increfs or not. If it doesnt then we need another +1

Reply:
I might have found what's happening. There is a chance this function is recursive and the returned val from "insert" was the argument (self) which we didn't want to decrement early to avoid ref == 0.
With the existing `nb::steal` we get protection for that case, and that being the case we shall not incref by ourselves.